pkg/usagestats/report.go

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package usagestats
+
+import (
+	"expvar"
+	"runtime"
+	"time"
+
+	prom "github.com/prometheus/prometheus/web/api/v1"
+
+	"github.com/grafana/mimir/pkg/util/version"
+)
+
+// Report is the JSON object sent to the stats server
+type Report struct {
+	// ClusterID is the unique Mimir cluster ID.
+	ClusterID string `json:"clusterID"`
+
+	// CreatedAt is when the cluster was created.
+	CreatedAt time.Time `json:"createdAt"`
+
+	// Interval is when the report was created (value is aligned across all replicas of the same Mimir cluster).
+	Interval time.Time `json:"interval"`
+
+	// IntervalPeriod is how frequently the report is sent, in seconds.
+	IntervalPeriod float64 `json:"intervalPeriod"`
+
+	// Target used to run Mimir.
+	Target string `json:"target"`
+
+	// Version holds information about the Mimir version.
+	Version prom.PrometheusVersion `json:"version"`
+
+	// Os is the operating system where Mimir is running.
+	Os string `json:"os"`
+
+	// Arch is the architecture where Mimir is running.
+	Arch string `json:"arch"`
+
+	// Edition is the Mimir edition ("oss" or "enterprise").
+	Edition string `json:"edition"`
+
+	// Metrics holds custom metrics tracked by Mimir. Can contain nested objects.
+	Metrics map[string]interface{} `json:"metrics"`
+}
+
+// buildReport builds the report to be sent to the stats server.
+func buildReport(seed ClusterSeed, reportAt time.Time, reportInterval time.Duration) Report {
+	var (
+		targetName  string
+		editionName string
+	)
+	if target := expvar.Get(statsPrefix + targetKey); target != nil {
+		if target, ok := target.(*expvar.String); ok {
+			targetName = target.Value()
+		}
+	}
+	if edition := expvar.Get(statsPrefix + editionKey); edition != nil {
+		if edition, ok := edition.(*expvar.String); ok {
+			editionName = edition.Value()
+		}
+	}
```

Comment:
`expvar.Get` returns a `expvar.Var` which is a `Stringer`, so why not just:

```suggestion
	if target := expvar.Get(statsPrefix + targetKey); target != nil {
		targetName = target.String()
	}
	if edition := expvar.Get(statsPrefix + editionKey); edition != nil {
		editionName = edition.String()
	}
```

Reply:
The `String()` implementation of `expvar` returns the JSON representation so a string is wrapped with `"`, which is not what we want here.

Reply:
<img src="https://user-images.githubusercontent.com/1511481/183395792-8c3ce96d-d8e4-47ed-85ac-9af98498a298.png" width="200">

Huh, okay.